stan/math/rev/functor/integrate_1d.hpp

```diff
@@ -59,6 +76,94 @@ inline double gradient_of_f(const F &f, const double &x, const double &xc,
   return gradient;
 }
 
+/**
+ * Return the integral of f from a to b to the given relative tolerance
+ *
+ * @tparam T_a type of first limit
+ * @tparam T_b type of second limit
+ * @tparam T_theta type of parameters
+ * @tparam T Type of f
+ *
+ * @param f the functor to integrate
+ * @param a lower limit of integration
+ * @param b upper limit of integration
+ * @param relative_tolerance relative tolerance passed to Boost quadrature
+ * @param[in, out] msgs the print stream for warning messages
+ * @param args additional arguments to pass to f
+ * @return numeric integral of function f
+ */
+template <typename F, typename T_a, typename T_b, typename... Args,
+          require_any_st_var<T_a, T_b, Args...> * = nullptr>
+inline return_type_t<T_a, T_b, Args...> integrate_1d_impl(
+    const F &f, const T_a &a, const T_b &b, double relative_tolerance,
+    std::ostream *msgs, const Args &... args) {
+  static const char *function = "integrate_1d";
+  check_less_or_equal(function, "lower limit", a, b);
+
+  double a_val = value_of(a);
+  double b_val = value_of(b);
+
+  if (a_val == b_val) {
+    if (is_inf(a_val)) {
+      throw_domain_error(function, "Integration endpoints are both", a_val, "",
+                         "");
+    }
+    return var(0.0);
+  } else {
+    std::tuple<decltype(value_of(args))...> args_val_tuple(value_of(args)...);
+
+    double integral = integrate(
+        [&](const auto &x, const auto &xc) {
+          return apply([&](auto &&... args) { return f(x, xc, msgs, args...); },
+                       args_val_tuple);
+        },
+        a_val, b_val, relative_tolerance);
+
+    size_t num_vars_ab = count_vars(a, b);
+    size_t num_vars_args = count_vars(args...);
+    vari **varis = ChainableStack::instance_->memalloc_.alloc_array<vari *>(
+        num_vars_ab + num_vars_args);
+    double *partials = ChainableStack::instance_->memalloc_.alloc_array<double>(
+        num_vars_ab + num_vars_args);
+    double *partials_ptr = partials;
+
+    save_varis(varis, a, b, args...);
+
+    for (size_t i = 0; i < num_vars_ab + num_vars_args; ++i) {
+      partials[i] = 0.0;
+    }
+
+    if (!is_inf(a) && is_var<T_a>::value) {
```

Comment:
[optional] Generally with short circuit logic operators you want the cheaper (in this context compile time known) thing as the first value.